File: sdks/python/src/devhub_sdk/_jsonrpc.py
from __future__ import annotations

from typing import Any, Mapping

from ._validation import ensure_json_object
from .exceptions import DevHubRpcException

_INT64_MIN = -(1 << 63)
_INT64_MAX = (1 << 63) - 1


def validate_response_envelope(root: Any, request_id: str) -> dict[str, Any]:
    """校验 JSON-RPC 响应并返回 `result` 对象。"""

    if not isinstance(root, dict):
        raise RuntimeError("JSON-RPC 响应根必须为对象。")
    if root.get("jsonrpc") != "2.0":
        raise RuntimeError("JSON-RPC 响应的 jsonrpc 版本非法。")

    has_result = "result" in root
    has_error = "error" in root and root.get("error") is not None
    if has_result == has_error:
        raise RuntimeError("JSON-RPC 响应必须且只能包含 result 或 error。")

    if has_error:
        error = root.get("error")
        if not isinstance(error, Mapping):
            raise RuntimeError("JSON-RPC error 对象非法。")
        code = error.get("code")
        message = error.get("message")
        if not isinstance(code, int) or isinstance(code, bool):
            raise RuntimeError("JSON-RPC error.code 非法。")
        if not isinstance(message, str) or not message:
            raise RuntimeError("JSON-RPC error.message 非法。")
        data = None
        if "data" in error:
            try:
                data = ensure_json_object(error["data"], "JSON-RPC error.data")
            except ValueError as exc:
                raise RuntimeError(str(exc)) from exc
        raise DevHubRpcException(
            code=code,
            message=message,
            data=data,
            request_id=request_id,
        )

    response_id = read_response_id(root)
    if response_id != request_id:
        raise RuntimeError("JSON-RPC 响应的 id 与请求不匹配。")

    result = root.get("result")
    if not isinstance(result, dict):
        raise RuntimeError("JSON-RPC result 必须为对象。")

    return result


def read_response_id(root: Mapping[str, Any]) -> str:
    """读取 JSON-RPC 响应 id。"""

    if "id" not in root:
        raise RuntimeError("JSON-RPC 响应缺少 id 字段。")
    return normalize_jsonrpc_id(root["id"], context="JSON-RPC 响应的 id")


def normalize_jsonrpc_id(value: Any, *, context: str) -> str:
    """读取并校验符合 DevHub 协议边界的 JSON-RPC id。"""

    if isinstance(value, str):
        return value
    if isinstance(value, bool):
        raise RuntimeError(f"{context} 类型非法。")
    if isinstance(value, int):
        if _INT64_MIN <= value <= _INT64_MAX:
            return str(value)
        raise RuntimeError(f"{context} 超出 Int64 范围。")
    if isinstance(value, float):
        raise RuntimeError(f"{context} 必须为 Int64 范围内整数。")
    raise RuntimeError(f"{context} 类型非法。")

File: sdks/javascript/src/jsonrpc.ts
import { DevHubRpcError } from "./errors.js";
import { ensureRecord, isRecord } from "./validation.js";
import { createRandomUuid } from "./web-crypto.js";

export interface ResponseEnvelope {
  requestId: string;
  result: Record<string, unknown>;
  error?: Record<string, unknown>;
}

export interface PendingRequest {
  promise: Promise<Record<string, unknown>>;
  resolve: (value: Record<string, unknown>) => void;
  reject: (reason?: unknown) => void;
  timeoutId?: NodeJS.Timeout;
}

export function createHttpRequestId(): string {
  return `req-${createRandomUuid().replace(/-/g, "")}`;
}

export function createWebSocketRequestId(): string {
  return `ws-${createRandomUuid().replace(/-/g, "")}`;
}

export function startTimeout(controller: AbortController, timeoutMs?: number): NodeJS.Timeout | undefined {
  if (!timeoutMs || timeoutMs <= 0) {
    return undefined;
  }

  return setTimeout(() => controller.abort(), timeoutMs);
}

export function validateResponseEnvelope(payload: unknown, requestId: string): Record<string, unknown> {
  const root = ensureRecord(payload, "JSON-RPC response");
  if (root.jsonrpc !== "2.0") {
    throw new Error("JSON-RPC response has an invalid jsonrpc version.");
  }

  const responseId = readResponseId(root, "JSON-RPC response");
  if (responseId !== requestId) {
    throw new Error("JSON-RPC response id does not match the request id.");
  }

  const hasResult = "result" in root;
  const hasError = "error" in root && root.error !== null && root.error !== undefined;
  if (hasResult === hasError) {
    throw new Error("JSON-RPC response must contain exactly one of result or error.");
  }

  if (hasError) {
    throw buildRpcError(root.error as unknown, requestId);
  }

  const result = root.result as unknown;
  if (!isRecord(result)) {
    throw new Error("JSON-RPC result must be an object.");
  }

  return result;
}

export function tryGetResponse(root: Record<string, unknown>): ResponseEnvelope | null {
  if (!("id" in root)) {
    return null;
  }

  const requestId = readResponseId(root, "WebSocket JSON-RPC response");
  const hasResult = "result" in root;
  const hasError = "error" in root && root.error !== null && root.error !== undefined;
  if (!hasResult && !hasError) {
    return null;
  }

  if (hasResult === hasError) {
    throw new Error("WebSocket JSON-RPC response must contain exactly one of result or error.");
  }

  if (hasResult) {
    const result = root.result as unknown;
    if (!isRecord(result)) {
      throw new Error("WebSocket JSON-RPC result must be an object.");
    }

    return {
      requestId,
      result
    };
  }

  const error = root.error as unknown;
  if (!isRecord(error)) {
    throw new Error("WebSocket JSON-RPC error must be an object.");
  }

  return {
    requestId,
    result: {},
    error
  };
}

export function tryGetEventParams(root: Record<string, unknown>): Record<string, unknown> | null {
  if (root.method !== "hub.event") {
    return null;
  }

  if ("id" in root) {
    throw new Error("hub.event must be delivered as a notification.");
  }

  if ("result" in root || ("error" in root && root.error !== null && root.error !== undefined)) {
    throw new Error("hub.event notification must not contain result or error.");
  }

  if (!("params" in root) || !isRecord(root.params)) {
    throw new Error("hub.event.params must be an object.");
  }

  return root.params as Record<string, unknown>;
}

export function validateIncomingEnvelope(root: Record<string, unknown>): void {
  if (root.jsonrpc !== "2.0") {
    throw new Error("WebSocket JSON-RPC message has an invalid jsonrpc version.");
  }
}

export function buildRpcError(payload: unknown, requestId: string): DevHubRpcError {
  if (!isRecord(payload)) {
    throw new Error("JSON-RPC error must be an object.");
  }

  const code = payload.code;
  if (!Number.isInteger(code)) {
    throw new Error("JSON-RPC error.code must be an integer.");
  }

  const message = payload.message;
  if (typeof message !== "string" || !message.trim()) {
    throw new Error("JSON-RPC error.message must be a non-empty string.");
  }

  const data = payload.data;
  if (data !== undefined && !isRecord(data)) {
    throw new Error("JSON-RPC error.data must be an object when present.");
  }

  return new DevHubRpcError({
    code: code as number,
    message,
    data,
    requestId
  });
}

export function readResponseId(root: Record<string, unknown>, location: string): string {
  if (!("id" in root)) {
    throw new Error(`${location} is missing id.`);
  }

  const id = root.id;
  if (typeof id === "string") {
    return id;
  }

  if (typeof id === "number") {
    return String(id);
  }

  throw new Error(`${location} has an invalid id type.`);
}

export function createPendingRequest(timeoutMs: number | undefined, onTimeout: () => void): PendingRequest {
  let resolve: (value: Record<string, unknown>) => void = () => {};
  let reject: (reason?: unknown) => void = () => {};

  const promise = new Promise<Record<string, unknown>>((resolveFn, rejectFn) => {
    resolve = resolveFn;
    reject = rejectFn;
  });

  let timeoutId: NodeJS.Timeout | undefined;
  if (timeoutMs && timeoutMs > 0) {
    timeoutId = setTimeout(() => {
      onTimeout();
      reject(new Error("WebSocket request timed out."));
    }, timeoutMs);
  }

  return {
    promise,
    resolve,
    reject,
    timeoutId
  };
}
